crates/matrix-sdk-base/src/client.rs

```diff
@@ -162,13 +162,30 @@ impl BaseClient {
     /// Clones the current base client to use the same crypto store but a
     /// different, in-memory store config, and resets transient state.
     #[cfg(feature = "e2e-encryption")]
-    pub fn clone_with_in_memory_state_store(&self) -> Self {
+    pub async fn clone_with_in_memory_state_store(&self) -> Result<Self> {
         let config = StoreConfig::new().state_store(MemoryStore::new());
         let config = config.crypto_store(self.crypto_store.clone());
 
-        let mut result = Self::with_store_config(config);
-        result.room_key_recipient_strategy = self.room_key_recipient_strategy.clone();
-        result
+        let (room_info_notable_update_sender, _room_info_notable_update_receiver) =
+            broadcast::channel(100);
+
+        let copy = Self {
+            store: Store::new(config.state_store),
+            event_cache_store: config.event_cache_store,
+            crypto_store: self.crypto_store.clone(),
+            olm_machine: self.olm_machine.clone(),
```

Comment:
Can you add a comment here, basically stating that it's fine to clone the current `olm_machine` which embeds the `self.crypto_store` because it's the same `crypto_store` in both? (I wonder if implementing the proper cross-process lock mechanism could allow us to get rid of `BaseClient::crypto_store` — i fathom it's there because we sometimes re-create the `crypto_store`, but that's bad state duplication :fearful:)

Reply:
It's rather there because we can create a `OlmMachine` only once we're logged in, i.e. we require a `UserId` and `DeviceId`, you could convert this into a `UnitializedOlmMachine` instead of having two optionals, this would make things clearer I would say.